File: jungol/jungol3836.py
import sys

# 빠른 입출력
input = sys.stdin.readline

def solve():
    line = input().split()
    if not line: return
    n = int(line[0])

    adj = [[] for _ in range(n + 1)]
    for _ in range(n - 1):
        u, v, c = map(int, input().split())
        w = 1 if c == 1 else -1
        adj[u].append((v, w))
        adj[v].append((u, w))

    removed = [False] * (n + 1)
    sz = [0] * (n + 1)
    offset = n
    # cnt[거리][0:내부휴식지없음, 1:내부휴식지있음]
    cnt = [[0, 0] for _ in range(2 * n + 1)]
    ans = 0

    def get_sz(root):
        stack = [(root, -1)]
        order = []
        while stack:
            u, p = stack.pop()
            order.append((u, p))
            sz[u] = 1
            for v, w in adj[u]:
                if v != p and not removed[v]:
                    stack.append((v, u))
        for u, p in reversed(order):
            if p != -1: sz[p] += sz[u]
        return sz[root]

    def get_centroid(u, total):
        curr, p = u, -1
        while True:
            heavy = -1
            for v, w in adj[curr]:
                if v != p and not removed[v] and sz[v] > total // 2:
                    heavy = v
                    break
            if heavy == -1: return curr
            p, curr = curr, heavy

    # 현재 경로상에 나타난 거리들의 빈도 (내부 휴식지 체크용)
    # seen_dist[0]은 무게 중심 C를 의미하므로 1로 시작
    seen_dist = [0] * (2 * n + 1)

    def collect_paths(start_node, start_w):
        # (현재노드, 부모, 누적거리, 백트래킹여부)
        stack = [(start_node, -1, start_w, False)]
        res = []
        seen_dist[0 + offset] = 1 # Centroid(거리 0) 표시
        
        while stack:
            u, p, d, backtrack = stack.pop()
            if backtrack:
                seen_dist[d + offset] -= 1
                continue
            
            # 1. 무게 중심 C를 중간 발판으로 쓸 수 있는가? 
            # (즉, 나를 제외한 조상 중에 나와 같은 누적거리가 있는가?)
            f1 = (seen_dist[d + offset] > 0)
            
            # 2. C를 제외하고 내부에 합이 0인 지점이 있는가?
            # (즉, C를 제외한 조상 중 누적거리가 0인 지점이 있는가?)
            f2 = (seen_dist[0 + offset] > 1)
            
            res.append((d, f1, f2))
            
            stack.append((u, p, d, True))
            seen_dist[d + offset] += 1
            for v, w in adj[u]:
                if v != p and not removed[v]:
                    stack.append((v, u, d + w, False))
        
        seen_dist[0 + offset] = 0 # 초기화
        return res

    def decompose(u):
        nonlocal ans
        total = get_sz(u)
        if total < 2: return
        
        c = get_centroid(u, total)
        removed[c] = True
        
        active_indices = []
        # 무게 중심을 지나는 경로 조합
        for v, w in adj[c]:
            if not removed[v]:
                subtree = collect_paths(v, w)
                for d, f1, f2 in subtree:
                    target = -d + offset
                    # 매칭되는 상대편 노드 찾기
                    if f1:
                        # 내가 이미 휴식지가 있다면(f1), 상대는 합만 0이면 됨
                        ans += cnt[target][0] + cnt[target][1]
                    else:
                        # 내가 없다면, 상대가 휴식지(f1)를 가지고 있어야 함
                        ans += cnt[target][1]
                    
                    # 무게 중심 C를 시작점으로 하는 경로 (C-u)
                    if d == 0 and f2:
                        ans += 1
                
                # 현재 서브트리 데이터를 카운트 배열에 업데이트
                for d, f1, f2 in subtree:
                    cnt[d + offset][1 if f1 else 0] += 1
                    active_indices.append(d + offset)

        # 사용한 부분만 0으로 리셋
        for idx in active_indices:
            cnt[idx][0] = cnt[idx][1] = 0
        
        for v, w in adj[c]:
            if not removed[v]:
                decompose(v)

    decompose(1)
    print(ans)

solve()

File: jungol/jungol3836.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

typedef long long ll;

const int MAXN = 100005;
const int OFFSET = 100000; // 음수 인덱스 방지용 오프셋

vector<pair<int, int>> adj[MAXN];
int sz[MAXN];
bool removed[MAXN];
int cnt[2 * MAXN][2]; // [거리][타입: 0은 중간0 없음, 1은 중간0 있음]
int seen[2 * MAXN];   // DFS 중 현재 경로에서 나타난 거리 빈도
int N;
ll ans = 0;

// 서브트리 크기 계산
void get_sz(int u, int p) {
    sz[u] = 1;
    for (auto& edge : adj[u]) {
        int v = edge.first;
        if (v != p && !removed[v]) {
            get_sz(v, u);
            sz[u] += sz[v];
        }
    }
}

// 무게 중심(Centroid) 찾기
int get_centroid(int u, int p, int total) {
    for (auto& edge : adj[u]) {
        int v = edge.first;
        if (v != p && !removed[v] && sz[v] > total / 2) {
            return get_centroid(v, u, total);
        }
    }
    return u;
}

struct NodeData {
    int dist;
    bool type;
};

// 현재 Centroid로부터의 거리와 중간에 합이 0이 되는 지점이 있는지 수집
void collect(int u, int p, int d, vector<NodeData>& nodes) {
    bool t = (seen[d + OFFSET] > 0);
    nodes.push_back({d, t});

    seen[d + OFFSET]++;
    for (auto& edge : adj[u]) {
        int v = edge.first;
        int w = (edge.second == 1 ? 1 : -1);
        if (v != p && !removed[v]) {
            collect(v, u, d + w, nodes);
        }
    }
    seen[d + OFFSET]--;
}

void decompose(int u) {
    get_sz(u, -1);
    if (sz[u] < 3 && u == 1 && N < 3) return; // 노드가 너무 적으면 종료
    int c = get_centroid(u, -1, sz[u]);
    removed[c] = true;

    vector<int> active_dists; // 리셋할 거리 목록

    for (auto& edge : adj[c]) {
        int v = edge.first;
        int w = (edge.second == 1 ? 1 : -1);
        if (!removed[v]) {
            vector<NodeData> subtree_nodes;
            collect(v, c, w, subtree_nodes);
            
            for (auto& node : subtree_nodes) {
                int target = -node.dist;
                if (node.dist == 0) {
                    // Centroid가 중간 지점인 경우 (u-C-v)
                    ans += cnt[OFFSET][0] + cnt[OFFSET][1];
                    // Centroid가 도달하기 전 중간 지점이 있었던 경우 (u-M-C)
                    if (node.type) ans++;
                } else {
                    // 일반적인 경우: u쪽이나 v쪽 서브트리에 중간 지점이 있어야 함
                    if (node.type) {
                        ans += cnt[target + OFFSET][0] + cnt[target + OFFSET][1];
                    } else {
                        ans += cnt[target + OFFSET][1];
                    }
                }
            }

            for (auto& node : subtree_nodes) {
                cnt[node.dist + OFFSET][node.type]++;
                active_dists.push_back(node.dist);
            }
        }
    }

    // 다음 Centroid를 위해 카운트 배열 초기화
    for (int d : active_dists) {
        cnt[d + OFFSET][0] = 0;
        cnt[d + OFFSET][1] = 0;
    }
    cnt[OFFSET][0] = 0;
    cnt[OFFSET][1] = 0;

    for (auto& edge : adj[c]) {
        if (!removed[edge.first]) {
            decompose(edge.first);
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    if (!(cin >> N)) return 0;
    for (int i = 0; i < N - 1; i++) {
        int u, v, c;
        cin >> u >> v >> c;
        adj[u].push_back({v, c});
        adj[v].push_back({u, c});
    }

    decompose(1);

    cout << ans << endl;

    return 0;
}

